### ISSUE TITLE
training in tf2 eager mode

### ISSUE THREAD

**Issue by mishooax** [jleinonen/downscaling-rnn-gan#7]
Hi @jleinonen - very nice work, thank you for sharing the code! I'm trying to use your GAN code in an atmospheric pollutant SR scenario. I have a question re. the following snippet in `layers.py`:

```
class SNDense(Dense):
    # ...
    def _get_trainable_var(self):
        if self._trainable_var is None:
            self._trainable_var = K.freezable_variable(self._trainable, name=self.name + "_trainable")
        return self._trainable_var

    def _get_training_value(self, training=None):
        if training is None:
            training = K.learning_phase()
        if isinstance(training, int):
            training = bool(training)
        if base_layer_utils.is_in_keras_graph():
            training = math_ops.logical_and(training, self._get_trainable_var())
        else:
            training = math_ops.logical_and(training, self.trainable)
        return training

    def call(self, inputs, training=None):
        training = self._get_training_value(training)
```

This fails to run if I enable eager mode (TF 2.2.0-gpu). Also - it doesn't work with more recent versions of TF (2.3+) - the `freezable_variable` is not present in the Keras backend anymore. I'm a bit puzzled as to the purpose of this code. Is it meant to enable / disable updates to the layer weights during the GAN training loop? Any idea how I can refactor this to work w/ newer TF versions? Thank you!

**Comment by jleinonen** [comment]
Hi @mishooax, I recall having various problems with getting the SNDense and SNConv layers to work properly and I'm not that surprised that it doesn't work well with another TF version. I haven't run the code in a while now and certainly not with the newer versions of TF. I'm glad you could get it running in non-eager mode, anyway.

I recently saw a neat-looking implementation of spectral normalization here: https://github.com/grohith327/simplegan. I haven't tried it myself yet, so I cannot guarantee that it works, but you might be able to replace the spectrally normalized layers with that implementation.

**Comment by mishooax** [comment]
Hi Jussi - thanks for replying! I was able to run the code in eager mode (TF 2.2) after replacing the SNDense and SNConv layers with conventional Dense and Conv layers, respectively. I'll look into the `simplegan` implementation that you referred me to - thanks for that. One further q: did you get a net benefit from using _both_ GP and spectral normalization in your WGAN architecture? The papers I've read - e.g., [http://proceedings.mlr.press/v97/kurach19a/kurach19a.pdf](http://proceedings.mlr.press/v97/kurach19a/kurach19a.pdf) - always seem to consider one or the other, but never in combination. Thanks.

**Comment by jleinonen** [comment]
For this project I honestly didn't try the different combinations - I had a previous GAN code where spectral normalization worked so I just went with it. Of course SN and GP try to impose similar constraints so maybe they aren't both necessary. So if it works for you without the SN layers I don't see anything necessarily wrong with that.

**Comment by mishooax** [comment]
Thanks, this makes sense. I'll close this issue now.
